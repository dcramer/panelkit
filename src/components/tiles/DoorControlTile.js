import React from "react";

import Tile from "../Tile";
import DoorControlModal from "../DoorControlModal";

export default class DoorControlTile extends Tile {
  onClick = () => {
    if (this.state.showModal) return;
    this.setState({ showModal: true });
  };

  closeModal = () => {
    this.setState({ showModal: false });
  };

  getIcon() {
    if (this.props.icon) return this.props.icon;
    return "door";
  }

  renderBody() {
    return (
      <React.Fragment>
        {super.renderBody()}
        <DoorControlModal
          hass={this.props.hass}
          name={this.renderLabel()}
          camera={this.props.camera}
          isOpen={this.state.showModal}
          onRequestClose={this.closeModal}
        />
      </React.Fragment>
    );
  }

  renderLabel() {
    return this.props.name || "Door Control";
  }
}
