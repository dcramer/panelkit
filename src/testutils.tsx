import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
