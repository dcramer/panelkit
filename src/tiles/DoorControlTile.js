import React from "react";
import PropTypes from "prop-types";

import Tile, { TileProps } from "./Tile";
import DoorControlModal from "../components/DoorControlModal";

export default class DoorControlTile extends Tile {
  static defaultIcon = "door";

  static PropTypes = {
    ...TileProps,
    camera: PropTypes.string.isRequired,
  };

  onTouch = () => {
    this.openModal();
  };

  renderModal({ ...props }) {
    console.log(this.props);
    return (
      <DoorControlModal
        title={this.props.title}
        camera={this.props.camera}
        {...props}
      />
    );
  }

  renderTitle() {
    return this.props.title || "Door Control";
  }
}
