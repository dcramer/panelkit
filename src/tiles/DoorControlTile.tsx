import React from "react";

import Tile, { TileProps, ModalParams } from "./Tile";
import DoorControlModal from "../modals/DoorControlModal";

type DoorControlTileProps = TileProps & {
  camera: string;
  title: string;
};

export default class DoorControlTile extends Tile<DoorControlTileProps> {
  static defaultIcon = "door";

  onTouch = async () => {
    this.openModal();
  };

  renderModal(params: ModalParams) {
    return (
      <DoorControlModal
        title={this.props.title}
        camera={this.props.camera}
        {...params}
      />
    );
  }

  renderTitle() {
    return this.props.title || "Door Control";
  }
}
