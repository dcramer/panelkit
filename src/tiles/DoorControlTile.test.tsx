import { renderTile } from "../testutils";
import DoorControlTile from "./DoorControlTile";

test("renders", async () => {
  const { container, getByText } = renderTile(
    DoorControlTile,
    {
      entity_id: "camera.test",
      state: "idle",
      attributes: {},
    },
    {
      entityId: undefined,
      camera: "camera.test",
      title: "Test Door Control",
    }
  );

  expect(getByText(/Test Door Control/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
