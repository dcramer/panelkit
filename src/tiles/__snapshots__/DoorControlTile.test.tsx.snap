// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile  touchable"
  style="cursor: pointer;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-body"
    >
      <span
        class="tile-icon"
      >
        <svg />
      </span>
    </div>
    <div
      class="tile-title"
    >
      Test Door Control
    </div>
  </div>
</div>
`;
